src/Elastic.Apm/Logging/IApmLoggingExtensions.cs

```diff
@@ -12,13 +13,87 @@
+	private static readonly ConcurrentDictionary<string, ScopedLogger> ScopedLoggers = new();
 
-	internal static ScopedLogger Scoped(this IApmLogger logger, string scope) => new(logger is ScopedLogger s ? s.Logger : logger, scope);
+	private static readonly ConditionalWeakTable<string, LogValuesFormatter> Formatters = new();
 
-	private static void DoLog(this IApmLogger logger, LogLevel level, string message, Exception e, params object[] args)
+	/// <summary>
+	/// Returns a new ScopedLogger.
+	/// </summary>
+	/// <remarks>This does not use the cashe of scoped loggers. Therefore it is strongly recommended
+	/// to use the scope based logging externsions such as <see cref="LogScopedError(IApmLogger?, ScopeName, string, object[])"/> instead.</remarks>
+	/// <param name="logger">An existing <see cref="IApmLogger"/>.</param>
+	/// <param name="scope">The name of the scope.</param>
+	/// <returns>A new scoped logger or <c>null</c> of the <see cref="IApmLogger"/> is <c>null</c>.</returns>
+	/// <exception cref="ArgumentException">Requires <paramref name="scope"/> to be non-null and non-empty.</exception>
+	internal static ScopedLogger? Scoped(this IApmLogger? logger, string scope)
+	{
+		if (string.IsNullOrEmpty(scope))
+			throw new ArgumentException("Scope is required to be non-null and non-empty.", nameof(scope));
+
+		if (logger is null)
+			return null;
+
+		// Ensure we don't allow creations of scoped loggers 'wrapping' other scoped loggers
+		return new(logger is ScopedLogger s ? s.Logger : logger, scope);
+	}
+
+	internal static void LogScopedDebug(this IApmLogger? logger, ScopeName scope, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Debug, message, null, args);
+
+	internal static void LogScopedTrace(this IApmLogger? logger, ScopeName scope, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Trace, message, null, args);
+
+	internal static void LogScopedTraceWithException(this IApmLogger? logger, ScopeName scope, Exception exception, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Trace, message, exception, args);
+
+	internal static void LogScopedWarning(this IApmLogger? logger, ScopeName scope, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Warning, message, null, args);
+
+	internal static void LogScopedWarningWithException(this IApmLogger? logger, ScopeName scope, Exception exception, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Warning, message, exception, args);
+
+	internal static void LogScopedError(this IApmLogger? logger, ScopeName scope, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Error, message, null, args);
+
+	internal static void LogScopedErrorWithException(this IApmLogger? logger, ScopeName scope, Exception exception, string message, params object?[]? args) =>
+		LogScoped(logger, scope, LogLevel.Error, message, exception, args);
```

Comment:
The idea behind the `?.` notation was to avoid evaluation of the log statement when the log level of the given line is not matching the current log level. That is because sometimes in e.g. debug logs we access things that may be expensive to get (some getters trigger crazy things), so we only wanted to access those when we really need them. By getting rid of the `?.` notation and using these methods we'd loose this. 

Not the best example, but e.g. here is one that shows this

In `Span.cs` `_logger.Trace()?.LogException` is replaced by `_logger.LogScopedTraceWithException`. I did a test and added back the `?.` notation before the `_logger.Trace()?.LogException` call to compare the result.
Below is the decompiled version to C#1 (IL is the real truth, but this is easier to read :) ):

<img width="2315" alt="Screenshot 2024-11-27 at 17 27 51" src="https://github.com/user-attachments/assets/cf483d60-abfa-492e-ba0b-e0c120fab9bd">

As you can see, it results in an if block, which does not run when the log level is not matching, so e.g. no getter is accessed on the HTTP context. Same for all the other things in similar scenarios - we fetch and concatenate a lot of things there that'd only want to evaluate when we really print those things.

Reply:
I've refactored back to a smaller phase one change to account for the above. Indeed we incurred some overhead for the log arguments with the changes, so we're now seeing an even bigger overall improvement.